internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavacProcessLogger.scala

```diff
@@ -20,43 +20,35 @@ import scala.sys.process.ProcessLogger
  * dump logs.
  *
  *
- * @param log  The logger where all input will go.
+ * @param log  The logger where all non-semantic messages will go.
  * @param reporter  A reporter for semantic Javac error messages.
  * @param cwd The current working directory of the Javac process, used when parsing Filenames.
  */
 final class JavacLogger(log: sbt.util.Logger, reporter: Reporter, cwd: File) extends ProcessLogger {
-  import scala.collection.mutable.ListBuffer
-  import sbt.util.Level.{ Info, Error, Value => LogLevel }
-
-  private val msgs: ListBuffer[(LogLevel, String)] = new ListBuffer()
+  private var out = ""
+  private var err = ""
 
   def out(s: => String): Unit =
-    synchronized { msgs += ((Info, s)); () }
+    synchronized { out += s }
 
   def err(s: => String): Unit =
-    synchronized { msgs += ((Error, s)); () }
+    synchronized { err += s }
 
   def buffer[T](f: => T): T = f
 
-  // Helper method to dump all semantic errors.
-  private def parseAndDumpSemanticErrors(): Unit = {
-    val input =
-      msgs collect {
-        case (Error, msg) => msg
-      } mkString "\n"
-    val parser = new JavaErrorParser(cwd)
-    parser.parseProblems(input, log).foreach(reporter.log)
-  }
-
-  def flush(exitCode: Int): Unit = {
-    parseAndDumpSemanticErrors()
-    // Here we only display things that wouldn't otherwise be output by the error reporter.
+  def flush(toolname: String, exitCode: Int): Unit = {
     // TODO - NOTES may not be displayed correctly!
-    msgs collect {
-      case (Info, msg) => msg
-    } foreach { msg =>
-      log.info(msg)
+    synchronized {
+      val parser = new JavaErrorParser(cwd)
+
+      parser.parseProblems(err, log).foreach(reporter.log(_))
+      out.split("\n").filter(_ != "").foreach(log.info(_))
```

Comment:
Are you purposely splitting info messages by newlines to send them independently to log.info?

Why not, instead of concat-ing info messages and then re-splitting, just collect the messages into a list buffer and then iterate that list into log.info?

Reply:
Hmm, for some reason I got the impression that we couldn't assume `out` would be called in neatly framed chunks, but indeed looking at the previous implementation again it appears that we in fact can. In that case a `ListBuffer` indeed makes more sense, I'll update accordingly!